pallets/compliance-manager/src/lib.rs

```diff
@@ -450,92 +451,85 @@ decl_event!(
-        condition: &'a Condition,
-        slot: &'a mut Option<Vec<TrustedIssuer>>,
-    ) -> &'a [TrustedIssuer] {
+    /// Returns the trusted issuers for the given `condition`.
+    /// If it doesn't have any, the asset's trusted issuers is returned.
+    fn get_trusted_issuers(condition: &Condition, ticker: &Ticker) -> Vec<TrustedIssuer> {
         if condition.issuers.is_empty() {
-            slot.get_or_insert_with(|| Self::trusted_claim_issuer(ticker))
+            Self::trusted_claim_issuer(ticker)
         } else {
-            &condition.issuers
+            condition.issuers.clone()
         }
     }
 
-    /// Fetches the proposition context for target `id` and specific `condition`.
-    /// Default trusted issuers, if fetched, are cached in `slot`.
-    fn fetch_context<'a>(
+    /// Returns all claims from the trusted issuers for the given identity.
+    fn fetch_identity_claims(
         id: IdentityId,
         ticker: &Ticker,
-        slot: &'a mut Option<Vec<TrustedIssuer>>,
-        condition: &'a Condition,
-    ) -> proposition::Context<impl 'a + Iterator<Item = Claim>> {
-        // Because of `-> impl Iterator`, we need to return a **single type** in each of the branches below.
-        // To do this, we use `Either<Either<MatchArm1, MatchArm2>, MatchArm3>`,
-        // equivalent to a 3-variant enum with iterators in each variant corresponding to the branches below.
-        // `Left(Left(arm1))`, `Left(Right(arm2))` and `Right(arm3)` correspond to arms 1, 2 and 3 respectively.
-        use either::Either::{Left, Right};
-
-        let claims = match &condition.condition_type {
-            ConditionType::IsPresent(claim) | ConditionType::IsAbsent(claim) => Left(Left(
-                Self::fetch_claims(id, claim, Self::issuers_for(ticker, condition, slot)),
-            )),
-            ConditionType::IsAnyOf(claims) | ConditionType::IsNoneOf(claims) => {
-                let issuers = Self::issuers_for(ticker, condition, slot);
-                Left(Right(claims.iter().flat_map(move |claim| {
-                    Self::fetch_claims(id, claim, issuers)
-                })))
+        condition: &Condition,
+    ) -> IdentityClaims {
+        let claims = {
+            match &condition.condition_type {
+                ConditionType::IsPresent(claim) | ConditionType::IsAbsent(claim) => {
+                    let trusted_issuers = Self::get_trusted_issuers(condition, ticker);
+                    Self::fetch_claims(id, claim, &trusted_issuers)
+                }
+                ConditionType::IsAnyOf(claims) | ConditionType::IsNoneOf(claims) => {
+                    let trusted_issuers = Self::get_trusted_issuers(condition, ticker);
+                    claims
+                        .iter()
+                        .flat_map(|claim| Self::fetch_claims(id, claim, &trusted_issuers))
+                        .collect()
+                }
+                ConditionType::IsIdentity(_) => Vec::new(),
```

Comment:
I think it would be much simpler to just eval the conditions here and not use the proposition logic.